near/mpc-recovery
# adding experiment with self calls

Adding a prototype of what the self call approach described in https://github.com/near/NEPs/issues/516#issuecomment-1854214737 would look like:
* `sign` function recursively calls itself until it runs out of gas
* `respond` function can add the requested signatures and allow `sign` to return the signature

To test this implementation:
* first compile the contract by `cargo build -p mpc-contract --target wasm32-unknown-unknown --release`
* Then start a local node by `nearup run localnet --num-shards 1 --binary-path <path-for-nearcore-binary>`
* Then create a new account on this localnet (assuming it is `mpc.node0` below)
* Deploy the contract `env NEAR_ENV=localnet near deploy mpc.node0 <path-to-contract-wasm> init '{"threshold": 1, "participants":{}}' --keyPath ~/.near-credentials/localnet/mpc.node0.json`
* call the contract `env NEAR_ENV=localnet near call mpc.node0 sign '{"payload":[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}' --accountId mpc.node0 --keyPath ~/.near-credentials/localnet/mpc.node0.json --gas 300000000000000`
Observe that it eventually returns a "exceeded prepaid gas" error after recursing 52 times.

Now to see that `respond` works. We can submit a signature by
```
env NEAR_ENV=localnet near call mpc.node0 respond '{"receipt_id":[102, 104, 122, 173, 248, 98, 189, 119, 108, 143, 193, 139, 142, 159, 142, 32, 8, 151, 20, 133, 110, 226, 51, 179, 144, 42, 89, 29, 13, 95, 41, 37], "big_r":"", "s":"something"}' --accountId mpc.node0 --keyPath ~/.near-credentials/localnet/mpc.node0.json
```
if there is a `sign` call running at the same time, it will terminate and return `something`.
